src/Streamly/Internal/Data/Stream/StreamD.hs

```diff
@@ -3844,3 +3849,21 @@ tapAsync f (Stream step1 state1) = Stream step TapInit
             Yield a s -> Yield a (TapDone s)
             Skip s    -> Skip (TapDone s)
             Stop      -> Stop
+
+-- XXX Is there room for improvement?
+-- | Take last 'n' elements from the stream and discard the rest.
+{-# INLINABLE lastN #-}
+lastN :: (Storable a, MonadIO m) => Int -> Fold m a (Array a)
+lastN n = Fold step initial done
+    where
+        step (Tuple3' rb rh i) a = do
+            rh1 <- liftIO $ RB.unsafeInsert rb rh a
+            return $ Tuple3' rb rh1 (i + 1)
+        initial = fmap (\(a, b) -> Tuple3' a b 0) $ liftIO $ RB.new n 
+        done (Tuple3' rb rh i) = do
+            arr <- liftIO $ A.newArray n
+            foldFunc i rh snoc' arr rb
+        snoc' b a = liftIO $ A.unsafeSnoc b a
```

Comment:
unsafeSnoc performs an array bound check, we do not need that here as we have already allocated an array of correct size, maybe that can help improve perf a bit.